website/blog/2023-04-04-presto-hudi-deep-dive.md

```diff
@@ -0,0 +1,110 @@
+5. [RealtimeCompactedRecordReader](https://github.com/apache/hudi/blob/master/hudi-hadoop-mr/src/main/java/org/apache/hudi/hadoop/realtime/RealtimeCompactedRecordReader.java)
+6. [HoodieMergedLogRecordScanner](https://github.com/apache/hudi/blob/master/hudi-common/src/main/java/org/apache/hudi/common/table/log/HoodieMergedLogRecordScanner.java)
+
+All the above classes are provided via `hudi-hadoop-mr` module in hudi repo, and we will see in some time how these classes are triggered.
+
+`BackgroundHiveSplitLoader` performs the task of loading splits and it in turn delegates the task to `StoragePartitionLoader.java` class. StoragePartitionLoader loads the partitions. After verifying the annotations and inputFormat class [here](https://github.com/prestodb/presto/blob/20e6b25e90828dd6db49ef89e7750bc9dc75d743/presto-hive/src/main/java/com/facebook/presto/hive/StoragePartitionLoader.java#L290), `inputFormat.getSplits()` is called. This is the first place where Hudi’s custom logic is called. Before diving deeper, let us have a look at the class hierarchies below - 
+
+![](/img/blog/2023-04-04-presto-hudi-deep-dive/input-format-hierarchy.png)
+
+*Figure 5: Input format hierarchy.*
+
+HoodieTableInputFormat is an abstract class and is implemented as below - 
+
+![](/img/blog/2023-04-04-presto-hudi-deep-dive/hudi-table-if-impl.png)
+*Figure 6: HoodieTableInputFormat implementations.*
+
+For `_rt` tables, input format is registered as `HoodieParquetRealtimeInputFormat`. Hence when `inputFormat.getSplits()` is called, it calls `HoodieParquetInputFormatBase.getSplits()` which delegates the call to `HoodieTableInputFormat` which is implemented as `HoodieMergeOnReadTableInputFormat.getSplits()`. This makes a call to `FileInputFormat.getSplits()`.
+
+Now refer to Figure 1, `listStatus()` is called which is overridden in `HoodieCopyOnWriteTableInputFormat.listStatus()`. This method in turn calls `listStatusForIncrementalMode()` and `listStatusForSnapshotMode()`. This will return `RealtimeFileStatus`. `RealtimeFileStatus` overrides `getPath()` method which returns `HoodieRealtimePath`. Please refer to the below figures -  
+
+![](/img/blog/2023-04-04-presto-hudi-deep-dive/list-status-mor.png)
+*Figure 7: listStatus() call for MoR table type returns RealtimeFileStatus.*
+
+![](/img/blog/2023-04-04-presto-hudi-deep-dive/make-split-mor.png)
+
+*Figure 8: makeSplit() internally called from getSplits() for MoR table type returns HoodieRealtimeFileSplit.*
+
+![](/img/blog/2023-04-04-presto-hudi-deep-dive/get-path-mor.png)
+
+*Figure 9: getPath() is overridden in RealtimeFileStatus and it returns HoodieRealtimePath.*
+
+Now that real time splits got generated, we need a way of reading the records from these special splits by merging the base and delta files. `GenericHiveRecordCursorProvider` calls createRecordReader which ultimately calls `inputFormat.getRecordReader()`. This returns `RealtimeCompactedRecordReader` by default. Please refer to the below figure - 
```

Comment:
Should we also explain the merging of records, or at the very least, why we're using the record reader as we need  custom reader that understand Hudi MOR format?

Reply:
Sure.